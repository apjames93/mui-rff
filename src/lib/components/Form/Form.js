/* eslint-disable max-len */
import React, { cloneElement } from 'react';
import PropTypes from 'prop-types';
import { Form as ReactFinalForm } from 'react-final-form';

export const RffWrapper = ({
  children,
  debug,
  initialValues,
  initialValuesEqual,
  keepDirtyOnReinitialize,
  onSubmit,
  subscription,
  validate,
  validateOnBlur,
  mutators,
}) => (
  <ReactFinalForm
    debug={debug}
    initialValues={initialValues}
    initialValuesEqual={initialValuesEqual}
    keepDirtyOnReinitialize={keepDirtyOnReinitialize}
    onSubmit={onSubmit}
    subscription={subscription}
    validate={validate}
    validateOnBlur={validateOnBlur}
    mutators={mutators}
  >
    {({ handleSubmit, submitting, form }) => (
      <form onSubmit={handleSubmit}>
        <>
          {Array.isArray(children) && children.map((child, key) => cloneElement(
            child,
            {
              key,
              submitting,
              form,
            },
          ))}

          {!Array.isArray(children) && cloneElement(
            children,
            {
              submitting,
              form,
            },
          )}
        </>
      </form>
    )}
  </ReactFinalForm>
);
export default RffWrapper;

// https://final-form.org/docs/react-final-form/types/FormProps
RffWrapper.propTypes = {
  /**
   * A render function that is given FormRenderProps, as well as any non-API props passed into the <Form/> component.
   */
  children: PropTypes.node.isRequired,
  /**
   * Function to call when the form is submitted. There are three possible ways to write an onSubmit function:
   */
  onSubmit: PropTypes.func.isRequired,

  /**
   * A callback for debugging that receives the form state and the states of all the fields. It's called on every state change. A typical thing to pass in might be console.log.
   */
  debug: PropTypes.func,
  /**
   * The initial values of your form. These will also be used to compare against the current values to calculate pristine and dirty.
   */
  initialValues: PropTypes.shape(),
  /**
   * A predicate to determine whether or not the initialValues prop has changed, i.e. to know if the form needs to be reinitialized with the new values. Useful for passing in a "deep equals" function if you need to. Defaults to "shallow equals".
   */
  initialValuesEqual: PropTypes.bool,
  /**
   * If true, only pristine values will be overwritten when initialize(newValues) is called. This can be useful for allowing a user to continue to edit a record while the record is being saved asynchronously, and the form is reinitialized to the saved values when the save is successful. Defaults to false.
   */
  keepDirtyOnReinitialize: PropTypes.bool,
  /**
   * An object of the parts of FormState to subscribe to. If a subscription is provided, the <Form/> will only rerender when those parts of form state change.
    If no subscription is provided, it will default to subscribing to all form state changes. i.e. <Form/> will rerender whenever any part of the form state changes.
   * ex : { [string]: boolean }
    */
  subscription: PropTypes.shape(),
  /**
   * ex: (values: FormValues) => Object | Promise<Object>
   * A whole-record validation function that takes all the values of the form and returns any validation errors. There are two possible ways to write a validate function:
   */
  validate: PropTypes.func,
  /**
   * If true, validation will happen on blur. If false, validation will happen on change. Defaults to false.
   */
  validateOnBlur: PropTypes.bool,
  /**
   * Optional.
   * Named mutator functions.
   */
  mutators: PropTypes.shape(),
};

RffWrapper.defaultProps = {
  debug: undefined,
  initialValues: {},
  initialValuesEqual: false,
  keepDirtyOnReinitialize: false,
  subscription: undefined,
  validate: undefined,
  validateOnBlur: false,
  mutators: undefined,
};
