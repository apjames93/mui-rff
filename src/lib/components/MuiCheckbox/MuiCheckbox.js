import React from 'react';
import PropTypes from 'prop-types';
import { Checkbox, FormControlLabel } from '@material-ui/core';
import MuiIcon from '../MUI/MuiIcon/MuiIcon';
import RFFField from '../RFFField/RFFField';

/**
 * MuiCheckbox
*/

export const MuiCheckbox = ({
  name,
  // mui
  disabled,
  checkedIcon,
  color,
  disableRipple,
  icon,
  indeterminate,
  size,
  label,
  passProps,
}) => (
  <RFFField
    type="checkbox"
    name={name}
    passProps={passProps}
  >
    <FormControlLabel
      control={(
        <Checkbox
          checkedIcon={checkedIcon?.iconName && <MuiIcon {...checkedIcon} />}
          color={color}
          disabled={disabled}
          disableRipple={disableRipple}
          icon={icon?.iconName && <MuiIcon {...icon} />}
          indeterminate={indeterminate}
          size={size}
        />
        )}
      label={label}
    />
  </RFFField>

);

export default MuiCheckbox;

MuiCheckbox.propTypes = {
  /**
   * props to pass to react final form field in call back have access to field as first arg
   */
  passProps: PropTypes.func,
  /**
   * The name of your field. Field values may be deeply nested using dot-and-bracket syntax.
   */
  name: PropTypes.string.isRequired,
  /**
 * The label content.
 */
  label: PropTypes.string,
  /**
 * material-ui. PROPS
 */

  /**
  * The icon to display when the component is checked.
  */
  checkedIcon: PropTypes.shape({
    /**
      * any icon from [material.io](https://material.io/resources/icons/?style=baseline)
      * ex star | alarm_off
      * */
    iconName: PropTypes.string.isRequired,
    /**
   * mui prop: "inherit", "primary", "secondary", "action", "error", "disabled"
   * The color of the component. It supports those theme colors that make sense for this component.
   * */
    color: PropTypes.string,
    /**
   * mui prop: 'default' | 'small'| 'inherit'| 'large'
   * The fontSize applied to the icon. Defaults to 24px, but can be configure to inherit font size.
   * */
    fontSize: PropTypes.string,
  }),
  /**
 * The color of the component. It supports those theme colors that make sense for this component.
 * mui prop: 'default' | 'primary' | 'secondary'
 */
  color: PropTypes.string,
  /**
 * If true, the checkbox will be disabled.
 */
  disabled: PropTypes.bool,
  /**
 * If true, the ripple effect will be disabled.
 */
  disableRipple: PropTypes.bool,

  /**
 * The icon to display when the component is unchecked.
 */
  icon: PropTypes.shape({
    /**
      * any icon from [material.io](https://material.io/resources/icons/?style=baseline)
      * ex star | alarm_off
      * */
    iconName: PropTypes.string.isRequired,
    /**
   * mui prop: "inherit", "primary", "secondary", "action", "error", "disabled"
   * The color of the component. It supports those theme colors that make sense for this component.
   * */
    color: PropTypes.string,
    /**
   * mui prop: 'default' | 'small'| 'inherit'| 'large'
   * The fontSize applied to the icon. Defaults to 24px, but can be configure to inherit font size.
   * */
    fontSize: PropTypes.string,
  }),
  /**
   * If true, the component appears indeterminate.
   * This does not set the native input element to indeterminate
   * due to inconsistent behavior across browsers.
   * However, we set a data-indeterminate attribute on the input.
   */
  indeterminate: PropTypes.bool,
  /**
     * The size of the checkbox. small is equivalent to the dense checkbox styling.
     * mui prop: 'medium'| 'small'
     */
  size: PropTypes.string,
};

MuiCheckbox.defaultProps = {
  passProps: field => ({
    id: field.input.name,
    value: field.input.value,
    onChange: field.input.onChange,
  }),
  // mui
  label: '',
  checkedIcon: null,
  color: 'secondary',
  disabled: false,
  disableRipple: false,
  icon: null,
  indeterminate: false,
  size: 'medium',
};
