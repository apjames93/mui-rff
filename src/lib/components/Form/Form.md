Form example:

```js noeditor
import Form from './Form';
import MuiSubmit from '../MuiSubmit/MuiSubmit';
import MuiInput from '../MuiInput/MuiInput';
import MuiSelect from '../MuiSelect/MuiSelect';
```

```js
import Form from './Form';
import MuiSubmit from '../MuiSubmit/MuiSubmit';
import MuiInput from '../MuiInput/MuiInput';
import MuiSelect from '../MuiSelect/MuiSelect';


const onSubmit = async (values) => {
  console.log('Input Form', values)
}

  <Form
    onSubmit={onSubmit}
  >
    <MuiInput
      name="thing"
      label="thing"
    />
    <MuiInput
      name="number"
      label="number"
      type="number"
    />
    <MuiInput
      name="date"
      label="date"
      type="date"
    />
    <MuiSelect
      name="stuff"
      options={[{ num: 'six' }, { num: 'nine' }]}
      displayKey="num"
    />
    <MuiSubmit />
  </Form>
```
