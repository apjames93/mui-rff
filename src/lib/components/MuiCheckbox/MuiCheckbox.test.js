import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { MuiCheckbox } from './MuiCheckbox';
import Form from '../Form/Form';

function setup() {
  const props = {
    name: 'cool',
  };
  const comp = shallow(<MuiCheckbox {...props} />);
  return { comp, props };
}

describe('<MuiCheckbox />', () => {
  it('renders Input', () => {
    const { comp } = setup();
    expect(comp).toBeDefined();
  });

  test('snapshot', () => {
    const { props } = setup();
    const tree = renderer.create(
      <Form onSubmit={() => true}>
        <MuiCheckbox {...props} />
      </Form>,
    );
    expect(tree).toMatchSnapshot();
  });

  it('should parse the checkbox to determine if checkbox is checked or not', () => {
    const { comp } = setup();
    const rfffield = comp.find('RFFField').first();
    const result = rfffield.prop('parse')([null]);
    expect(result).toEqual(true);
  });
});
