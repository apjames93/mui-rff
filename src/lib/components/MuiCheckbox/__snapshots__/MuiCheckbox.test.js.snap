// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MuiCheckbox /> snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <label
    className="MuiFormControlLabel-root"
    field={
      Object {
        "input": Object {
          "checked": false,
          "name": "cool",
          "onBlur": [Function],
          "onChange": [Function],
          "onFocus": [Function],
          "type": "checkbox",
          "value": null,
        },
        "meta": Object {
          "active": false,
          "data": Object {},
          "dirty": false,
          "dirtySinceLastSubmit": false,
          "error": undefined,
          "initial": false,
          "invalid": false,
          "length": undefined,
          "modified": false,
          "modifiedSinceLastSubmit": false,
          "pristine": true,
          "submitError": undefined,
          "submitFailed": false,
          "submitSucceeded": false,
          "submitting": false,
          "touched": false,
          "valid": true,
          "validating": false,
          "visited": false,
        },
      }
    }
    type="checkbox"
  >
    <span
      aria-disabled={false}
      className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={null}
    >
      <span
        className="MuiIconButton-label"
      >
        <input
          className="PrivateSwitchBase-input-4"
          data-indeterminate={false}
          disabled={false}
          name="cool"
          onChange={[Function]}
          type="checkbox"
          value={false}
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
      </span>
    </span>
    <span
      className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
    >
      
    </span>
  </label>
</form>
`;
